h5netcdf/core.py

```diff
@@ -125,28 +132,59 @@ def shuffle(self):
 
 NOT_A_VARIABLE = b'This is a netCDF dimension but not a netCDF variable.'
 
+class _object_names(Mapping):
+    def __init__(self, parent, object_cls):
+        self._parent = parent
+        self._object_cls = object_cls
+        self._objects = set()
```

Comment:
It occurs to me that `_objects` needs to be ordered, because objects in HDF5/netCDF files have an order which we should iterate over properly.

Python doesn't have an OrderedSet in the standard library, so our best choice might be to use an OrderedDict for `_objects` instead. This suggests an alternative strategy for this class:

- Save unloaded objects in `_objects` as `None`.
- When calling `__getitem__` on an object that is None, replace it with real object.